dbus-wrapper/swupdd-main.c

```diff
@@ -0,0 +1,985 @@
+	if (r < 0) {
+		sd_bus_error_set_errnof(error, -r, "Can't read value of '%s'", optname);
+		return r;
+	}
+	r = sd_bus_message_exit_container(m);
+	if (r < 0) {
+		sd_bus_error_set_errnof(error, -r, "Can't exit variant container of '%s'", optname);
+		return r;
+	}
+
+	if (value) {
+		char *option = NULL;
+
+		r = asprintf(&option, "--%s", optname);
+		if (r < 0) {
+			sd_bus_error_set_errnof(error, ENOMEM, "Can't allocate memory for '%s'", optname);
+			return -ENOMEM;
+		}
+		*strlist = list_append_data(*strlist, option);
+	}
+	return 0;
+}
+
+static int bus_message_read_option_int(sd_bus_message *m,
+				       const char *optname,
+				       struct list **strlist,
+				       sd_bus_error *error)
+{
+	char *option = NULL;
+	char *value_str = NULL;
+	int value;
+	int r;
+
+	r = sd_bus_message_enter_container(m, SD_BUS_TYPE_VARIANT, "i");
+	if (r < 0) {
+		sd_bus_error_set_errnof(error, -r, "Failed to enter variant container of '%s'", optname);
+		return r;
+	}
+	r = sd_bus_message_read(m, "i", &value);
+	if (r < 0) {
+		sd_bus_error_set_errnof(error, -r, "Can't read value of '%s'", optname);
+		return r;
+	}
+	r = sd_bus_message_exit_container(m);
+	if (r < 0) {
+		sd_bus_error_set_errnof(error, -r, "Can't exit variant container of '%s'", optname);
+		return r;
+	}
+
+	if (asprintf(&option, "--%s", optname) < 0) {
+		sd_bus_error_set_errnof(error, ENOMEM, "Can't allocate memory for '%s'", optname);
+		return -ENOMEM;
+	}
+	*strlist = list_append_data(*strlist, option);
+
+	if (asprintf(&value_str, "%i", value) < 0) {
+		sd_bus_error_set_errnof(error, ENOMEM, "Can't allocate memory for '%i'", value);
+		return -ENOMEM;
+	}
+	*strlist = list_append_data(*strlist, value_str);
+
+	return 0;
+}
+
+static int bus_message_read_options(sd_bus_message *m,
+	                            char const * const opts_str[],
+	                            char const * const opts_bool[],
+				    char const * const opts_int[],
+				    struct list **strlist,
+				    sd_bus_error *error)
+{
+	int r;
+
+	r = sd_bus_message_enter_container(m, SD_BUS_TYPE_ARRAY, "{sv}");
+	if (r < 0) {
+		sd_bus_error_set_errnof(error, -r, "Failed to enter options container");
+		return r;
+	}
+
+	while ((r = sd_bus_message_enter_container(m, SD_BUS_TYPE_DICT_ENTRY, "sv")) > 0) {
+                const char *argname;
+
+                r = sd_bus_message_read(m, "s", &argname);
+                if (r < 0) {
+			sd_bus_error_set_errnof(error, -r, "Can't read option name");
+                        return r;
+		}
+		if (is_in_array(argname, opts_str)) {
+			r = bus_message_read_option_string(m, argname, strlist, error);
+			if (r < 0) {
+				return r;
+			}
+		} else if (is_in_array(argname, opts_bool)) {
+			r = bus_message_read_option_bool(m, argname, strlist, error);
+			if (r < 0) {
+				return r;
+			}
+		} else if (is_in_array(argname, opts_int)) {
+			r = bus_message_read_option_int(m, argname, strlist, error);
+			if (r < 0) {
+				return r;
+			}
+		} else {
```

Comment:
What is this else case handling exactly? Should this be treated as an input error instead of skipped?